File: src/Atlasd/Battlenet/Account.cs
﻿using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.Net;
using System.Text;
using System.Text.Json;

namespace Atlasd.Battlenet
{
    class Account
    {
        public const string AccountCreatedKey = "System\\Account Created";
        public const string FailedLogonsKey = "System\\Total Failed Logons";
        public const string FlagsKey = "System\\Flags";
        public const string FriendsKey = "System\\Friends";
        public const string IPAddressKey = "System\\IP";
        public const string LastLogoffKey = "System\\Last Logoff";
        public const string LastLogonKey = "System\\Last Logon";
        public const string PasswordKey = "System\\Password Digest";
        public const string PortKey = "System\\Port";
        public const string ProfileAgeKey = "profile\\age";
        public const string ProfileDescriptionKey = "profile\\description";
        public const string ProfileLocationKey = "profile\\location";
        public const string ProfileSexKey = "profile\\sex";
        public const string TimeLoggedKey = "System\\Time Logged";
        public const string UsernameKey = "System\\Username";

        public const string Alphanumeric = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        public const string Punctuation = "`~!$%^&*()-_=+[{]}\\|;:'\",<.>/?";

        public enum CreateStatus : UInt32
        {
            Success = 0,
            UsernameTooShort = 1,
            UsernameInvalidChars = 2,
            UsernameBannedWord = 3,
            AccountExists = 4,
            LastCreateInProgress = 5,
            UsernameShortAlphanumeric = 6,
            UsernameAdjacentPunctuation = 7,
            UsernameTooManyPunctuation = 8,
        };

        [Flags]
        public enum Flags : UInt32
        {
            None = 0x00,
            Employee = 0x01,
            ChannelOp = 0x02,
            Speaker = 0x04,
            Admin = 0x08,
            NoUDP = 0x10,
            Squelched = 0x20,
            Guest = 0x40,
            Closed = 0x80,
        };

        public List<AccountKeyValue> Userdata { get; protected set; }

        private Account()
        {
            Userdata = new List<AccountKeyValue>()
            {
                { new AccountKeyValue(AccountCreatedKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.ReadOnly) },
                { new AccountKeyValue(FailedLogonsKey, (long)0, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(FlagsKey, Flags.None, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(FriendsKey, new List<string>() { "Carl" }, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(IPAddressKey, IPAddress.Any, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(LastLogoffKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(LastLogonKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(PortKey, 0, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(PasswordKey, new byte[0], AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(ProfileAgeKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner) },
                { new AccountKeyValue(ProfileDescriptionKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner) },
                { new AccountKeyValue(ProfileLocationKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner) },
                { new AccountKeyValue(ProfileSexKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner) },
                { new AccountKeyValue(TimeLoggedKey, (long)0, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal) },
                { new AccountKeyValue(UsernameKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Internal) },
            };
        }

        public bool ContainsKey(string key)
        {
            var keyL = key.ToLower();

            lock (Userdata)
            {
                foreach (var kv in Userdata)
                {
                    if (kv.Key.ToLower() == keyL) return true;
                }
            }

            return false;
        }

        public bool Get(byte[] key, out dynamic value)
        {
            return Get(Encoding.UTF8.GetString(key), out value);
        }

        public bool Get(string key, out dynamic value)
        {
            var keyL = key.ToLower();

            lock (Userdata)
            {
                foreach (var kv in Userdata)
                {
                    if (kv.Key.ToLower() == keyL)
                    {
                        value = kv;
                        return true;
                    }
                }
            }

            value = null;
            return false;
        }

        public dynamic Get(string key, dynamic onKeyNotFound = null)
        {
            if (!Get(key, out dynamic value))
            {
                return onKeyNotFound;
            }

            if (value == null) return onKeyNotFound;
            if (!(value is AccountKeyValue)) return value;

            return ((AccountKeyValue)value).Value;
        }

        public void Set(string key, dynamic value)
        {
            var keyL = key.ToLower();

            lock (Userdata)
            {
                foreach (var kv in Userdata)
                {
                    if (kv.Key.ToLower() == keyL)
                    {
                        kv.Value = value;
                        return;
                    }
                }
            }

            throw new KeyNotFoundException(key);
        }

        public static CreateStatus TryCreate(string username, byte[] passwordHash, out Account account)
        {
            account = null;

            Settings.State.RootElement.TryGetProperty("account", out var accountJson);
            accountJson.TryGetProperty("auto_admin", out var autoAdminJson);
            accountJson.TryGetProperty("disallow_words", out var disallowWordsJson);
            accountJson.TryGetProperty("max_adjacent_punctuation", out var maxAdjacentPunctuationJson);
            accountJson.TryGetProperty("max_length", out var maxLengthJson);
            accountJson.TryGetProperty("max_punctuation", out var maxPunctuationJson);
            accountJson.TryGetProperty("min_alphanumeric", out var minAlphanumericJson);
            accountJson.TryGetProperty("min_length", out var minLengthJson);

            if (!disallowWordsJson.ValueKind.HasFlag(JsonValueKind.Array))
            {
                throw new NotSupportedException("Setting [account] -> [disallow_words] is not an array; check value");
            }

            bool autoAdmin = false;
            if (autoAdminJson.ValueKind == JsonValueKind.Array)
            {
                foreach (var nameJson in autoAdminJson.EnumerateArray())
                {
                    var name = nameJson.GetString();
                    if (username.ToLower() == name.ToLower())
                    {
                        autoAdmin = true;
                        break;
                    }
                }
            }
            else if (autoAdminJson.ValueKind == JsonValueKind.String)
            {
                autoAdmin = username.ToLower() == autoAdminJson.GetString().ToLower();
            }
            else if (autoAdminJson.ValueKind == JsonValueKind.True || autoAdminJson.ValueKind == JsonValueKind.False)
            {
                autoAdmin = autoAdminJson.GetBoolean() && Common.AccountsDb.Count == 0;
            }
            else
            {
                throw new NotSupportedException("Setting [account] -> [auto_admin] is not an array, string, or boolean; check value");
            }

            var bannedWords = disallowWordsJson;
            var maximumAdjacentPunctuation = maxAdjacentPunctuationJson.GetUInt32();
            var maximumPunctuation = maxPunctuationJson.GetUInt32();
            var maximumUsernameSize = maxLengthJson.GetUInt32();
            var minimumAlphanumericSize = minAlphanumericJson.GetUInt32();
            var minimumUsernameSize = minLengthJson.GetUInt32();

            lock (Common.AccountsProcessing)
            {
                if (Common.AccountsProcessing.Contains(username))
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Still processing new account request...");
                    return CreateStatus.LastCreateInProgress;
                }
                Common.AccountsProcessing.Add(username);
            }

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Processing new account request...");

            if (username.Length < minimumUsernameSize)
                return CreateStatus.UsernameTooShort;

            if (username.Length > maximumUsernameSize)
                return CreateStatus.UsernameShortAlphanumeric;

            uint total_alphanumeric = 0;
            uint total_punctuation = 0;
            uint adjacent_punctuation = 0;
            char last_c = (char)0;

            foreach (var c in username)
            {
                if (!Alphanumeric.Contains(c) && !Punctuation.Contains(c))
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Requested username contains invalid characters");
                    return CreateStatus.UsernameInvalidChars;
                }

                if (Alphanumeric.Contains(c))
                    total_alphanumeric++;

                if (Punctuation.Contains(c))
                {
                    total_punctuation++;

                    if (last_c != 0 && Punctuation.Contains(last_c))
                        adjacent_punctuation++;
                }

                if (total_punctuation > maximumPunctuation)
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{username}] contains too many punctuation");
                    return CreateStatus.UsernameTooManyPunctuation;
                }

                if (adjacent_punctuation > maximumAdjacentPunctuation)
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{username}] contains too many adjacent punctuation");
                    return CreateStatus.UsernameAdjacentPunctuation;
                }

                last_c = c;
            }

            if (total_alphanumeric < minimumAlphanumericSize)
            {
                Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{username}] is too short or contains too few alphanumeric characters");
                return CreateStatus.UsernameShortAlphanumeric;
            }

            foreach (var word in bannedWords.EnumerateArray())
            {
                if (username.ToLower().Contains(word.GetString().ToLower()))
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{username}] contains a banned word or phrase");
                    return CreateStatus.UsernameBannedWord;
                }
            }

            lock (Common.AccountsDb)
            {
                if (Common.AccountsDb.ContainsKey(username))
                {
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{username}] already exists");
                    return CreateStatus.AccountExists;
                }

                account = new Account();

                account.Set(Account.UsernameKey, username);
                account.Set(Account.PasswordKey, passwordHash);

                account.Set(Account.FlagsKey, autoAdmin ? (Account.Flags.Employee | Account.Flags.Admin) : Account.Flags.None);

                Common.AccountsDb.Add(username, account);
            }

            lock (Common.AccountsProcessing)
            {
                Common.AccountsProcessing.Remove(username);
            }

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Created new account [{username}]");
            return CreateStatus.Success;
        }
    }
}

File: src/AtlasV/Battlenet/BNET/Account.vb
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.Net
Imports System.Runtime.InteropServices
Imports System.Text
Imports System.Text.Json

Namespace AtlasV.Battlenet
    Class Account
        Public Const AccountCreatedKey As String = "System\Account Created"
        Public Const FailedLogonsKey As String = "System\Total Failed Logons"
        Public Const FlagsKey As String = "System\Flags"
        Public Const FriendsKey As String = "System\Friends"
        Public Const IPAddressKey As String = "System\IP"
        Public Const LastLogoffKey As String = "System\Last Logoff"
        Public Const LastLogonKey As String = "System\Last Logon"
        Public Const PasswordKey As String = "System\Password Digest"
        Public Const PortKey As String = "System\Port"
        Public Const ProfileAgeKey As String = "profile\age"
        Public Const ProfileDescriptionKey As String = "profile\description"
        Public Const ProfileLocationKey As String = "profile\location"
        Public Const ProfileSexKey As String = "profile\sex"
        Public Const TimeLoggedKey As String = "System\Time Logged"
        Public Const UsernameKey As String = "System\Username"
        Public Const Alphanumeric As String = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
        Public Const Punctuation As String = "`~!$%^&*()-_=+[{]}\|;:'"",<.>/?"

        Public Enum CreateStatus As UInt32
            Success = 0
            UsernameTooShort = 1
            UsernameInvalidChars = 2
            UsernameBannedWord = 3
            AccountExists = 4
            LastCreateInProgress = 5
            UsernameShortAlphanumeric = 6
            UsernameAdjacentPunctuation = 7
            UsernameTooManyPunctuation = 8
        End Enum

        <Flags>
        Public Enum Flags As UInt32
            None = &H0
            Employee = &H1
            ChannelOp = &H2
            Speaker = &H4
            Admin = &H8
            NoUDP = &H10
            Squelched = &H20
            Guest = &H40
            Closed = &H80
        End Enum

        Public Property Userdata As List(Of AccountKeyValue)

        Private Sub New()
            Userdata = New List(Of AccountKeyValue)() From {
                {New AccountKeyValue(AccountCreatedKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.[ReadOnly])},
                {New AccountKeyValue(FailedLogonsKey, CLng(0), AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(FlagsKey, Flags.None, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(FriendsKey, New List(Of Byte())(), AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(IPAddressKey, IPAddress.Any, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(LastLogoffKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(LastLogonKey, DateTime.Now, AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(PortKey, 0, AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(PasswordKey, Array.Empty(Of Byte)(), AccountKeyValue.ReadLevel.Internal, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(ProfileAgeKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner)},
                {New AccountKeyValue(ProfileDescriptionKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner)},
                {New AccountKeyValue(ProfileLocationKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner)},
                {New AccountKeyValue(ProfileSexKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Owner)},
                {New AccountKeyValue(TimeLoggedKey, CLng(0), AccountKeyValue.ReadLevel.Owner, AccountKeyValue.WriteLevel.Internal)},
                {New AccountKeyValue(UsernameKey, "", AccountKeyValue.ReadLevel.Any, AccountKeyValue.WriteLevel.Internal)}
            }
        End Sub

        Public Function ContainsKey(ByVal key As String) As Boolean
            Dim keyL = key.ToLower()

            SyncLock Userdata

                For Each kv In Userdata
                    If kv.Key.ToLower() = keyL Then Return True
                Next
            End SyncLock

            Return False
        End Function

        Public Function [Get](key() As Byte, ByRef value As Object) As Boolean
            Return [Get](Encoding.UTF8.GetString(key), value)
        End Function

        Public Function [Get](key As String, ByRef value As Object) As Boolean
            Dim keyL = key.ToLower()

            value = Nothing
            SyncLock Userdata
                For Each kv In Userdata
                    If kv.Key.ToLower() = keyL Then
                        value = kv
                        Exit For
                    End If
                Next
                If value Is Nothing Then Return False
                Return True
            End SyncLock

        End Function

        'VB VooDoo
#Disable Warning IDE0060 ' Remove unused parameter
        Public Function [Get](key As String, Optional onKeyNotFound As Object = Nothing, Optional IsType As Boolean = False) As Object
#Enable Warning IDE0060 ' Remove unused parameter
            Dim value As Object = Nothing

            If Not [Get](key, value) Then
                Return onKeyNotFound
            End If

            If value Is Nothing Then Return onKeyNotFound
            If Not (TypeOf value Is AccountKeyValue) Then Return value
            Return (CType(value, AccountKeyValue)).Value
        End Function

        Public Sub [Set](ByVal key As String, ByVal value As Object)
            Dim keyL = key.ToLower()

            SyncLock Userdata

                For Each kv In Userdata

                    If kv.Key.ToLower() = keyL Then
                        kv.Value = value
                        Return
                    End If
                Next
            End SyncLock

            Throw New KeyNotFoundException(key)
        End Sub

        Public Shared Function TryCreate(ByVal varUsername As String, ByVal varPasswordHash As Byte(), <Out> ByRef varAccount As Account) As CreateStatus
            varAccount = Nothing
            Dim accountJson As JsonElement = Nothing
            Settings.State.RootElement.TryGetProperty("account", accountJson)
            Dim autoAdminJson = Nothing
            accountJson.TryGetProperty("auto_admin", autoAdminJson)
            Dim disallowWordsJson = Nothing
            accountJson.TryGetProperty("disallow_words", disallowWordsJson)
            Dim maxAdjacentPunctuationJson = Nothing
            accountJson.TryGetProperty("max_adjacent_punctuation", maxAdjacentPunctuationJson)
            Dim maxLengthJson = Nothing
            accountJson.TryGetProperty("max_length", maxLengthJson)
            Dim maxPunctuationJson = Nothing
            accountJson.TryGetProperty("max_punctuation", maxPunctuationJson)
            Dim minAlphanumericJson = Nothing
            accountJson.TryGetProperty("min_alphanumeric", minAlphanumericJson)
            Dim minLengthJson = Nothing
            accountJson.TryGetProperty("min_length", minLengthJson)

            If Not disallowWordsJson.ValueKind.HasFlag(JsonValueKind.Array) Then
                Throw New NotSupportedException("Setting [account] -> [disallow_words] is not an array; check value")
            End If

            Dim autoAdmin As Boolean = False

            If autoAdminJson.ValueKind = JsonValueKind.Array Then

                For Each nameJson In autoAdminJson.EnumerateArray()
                    Dim name = nameJson.GetString()

                    If varUsername.ToLower() = name.ToLower() Then
                        autoAdmin = True
                        Exit For
                    End If
                Next
            ElseIf autoAdminJson.ValueKind = JsonValueKind.String Then
                autoAdmin = varUsername.ToLower() = autoAdminJson.GetString().ToLower()
            ElseIf autoAdminJson.ValueKind = JsonValueKind.[True] OrElse autoAdminJson.ValueKind = JsonValueKind.[False] Then
                autoAdmin = autoAdminJson.GetBoolean() AndAlso Common.AccountsDb.Count = 0
            Else
                Throw New NotSupportedException("Setting [account] -> [auto_admin] is not an array, string, or boolean; check value")
            End If

            Dim bannedWords As JsonElement = disallowWordsJson
            Dim maximumAdjacentPunctuation = maxAdjacentPunctuationJson.GetUInt32()
            Dim maximumPunctuation = maxPunctuationJson.GetUInt32()
            Dim maximumUsernameSize = maxLengthJson.GetUInt32()
            Dim minimumAlphanumericSize = minAlphanumericJson.GetUInt32()
            Dim minimumUsernameSize = minLengthJson.GetUInt32()

            SyncLock Common.AccountsProcessing

                If Common.AccountsProcessing.Contains(varUsername) Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Still processing new account request...")
                    Return CreateStatus.LastCreateInProgress
                End If

                Common.AccountsProcessing.Add(varUsername)
            End SyncLock

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Processing new account request...")
            If varUsername.Length < minimumUsernameSize Then Return CreateStatus.UsernameTooShort
            If varUsername.Length > maximumUsernameSize Then Return CreateStatus.UsernameShortAlphanumeric
            Dim total_alphanumeric As UInteger = 0
            Dim total_punctuation As UInteger = 0
            Dim adjacent_punctuation As UInteger = 0
            Dim last_c As Char = ChrW(0)

            For Each c In varUsername

                If Not Alphanumeric.Contains(c) AndAlso Not Punctuation.Contains(c) Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, "Requested username contains invalid characters")
                    Return CreateStatus.UsernameInvalidChars
                End If

                If Alphanumeric.Contains(c) Then total_alphanumeric += 1

                If Punctuation.Contains(c) Then
                    total_punctuation += 1
                    If last_c <> Chr(0) AndAlso Punctuation.Contains(last_c) Then adjacent_punctuation += 1
                End If

                If total_punctuation > maximumPunctuation Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] contains too many punctuation")
                    Return CreateStatus.UsernameTooManyPunctuation
                End If

                If adjacent_punctuation > maximumAdjacentPunctuation Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] contains too many adjacent punctuation")
                    Return CreateStatus.UsernameAdjacentPunctuation
                End If

                last_c = c
            Next

            If total_alphanumeric < minimumAlphanumericSize Then
                Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] is too short or contains too few alphanumeric characters")
                Return CreateStatus.UsernameShortAlphanumeric
            End If

            'This is not here to destroy your current "BannedWords" there is an actual banned word list for
            'accounts, also titles but that one was characters only. So pulled out the curse words that
            'exist in the profanity listing, since this takes care of those words as well as the main missing 
            'profane words.
            If AtlasV.Battlenet.Protocols.Game.ProfanityFilter.ContainsProfane(varUsername) Then
                Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] contains Profanity.")
                Return CreateStatus.UsernameBannedWord
            End If

            For Each word In bannedWords.EnumerateArray()
                If varUsername.ToLower().Contains(word.GetString().ToLower()) Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] contains a banned word or phrase")
                    Return CreateStatus.UsernameBannedWord
                End If
            Next

            SyncLock Common.AccountsDb

                If Common.AccountsDb.ContainsKey(varUsername) Then
                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Requested username [{varUsername}] already exists")
                    Return CreateStatus.AccountExists
                End If

                varAccount = New Account()
                varAccount.[Set](Account.UsernameKey, varUsername)
                varAccount.[Set](Account.PasswordKey, varPasswordHash)
                varAccount.[Set](Account.FlagsKey, If(autoAdmin, (Account.Flags.Employee Or Account.Flags.Admin), Account.Flags.None))
                Common.AccountsDb.Add(varUsername, varAccount)
            End SyncLock

            SyncLock Common.AccountsProcessing
                Common.AccountsProcessing.Remove(varUsername)
            End SyncLock

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Account, $"Created new account [{varUsername}]")
            Return CreateStatus.Success
        End Function
    End Class
End Namespace
